router2023/main.md

```diff
@@ -0,0 +1,454 @@
+# Nixos based router in 2023
+
+After spending a few months with NixOS as my daily driver and successfully fixing most of the issues,
+that I encountered (unknown to the happy people who use a user-friendly distro like ubuntu),
+
+I felt empowered and free now when I can put everything into a VCS.
+
+I decided to continue nixification even further. But what else can you nixify after switching your main operating system to NixOS?
+
+Look around, I bet you will find a lot of devices that still are not running NixOS :) One strong contender for that was my old router.
+
+Why I chose to ~break~ nixify my router:
+
+- It is a device of critical importance and it should be up-to-date due to security reasons, however, I have no idea if regular vendors are shipping any updates to the devices that they have already sold. On the other hand, keeping it up-to-date using nix will be trivial.
+- It can serve as a vpn server, provided that your firmware allows for that. That is not an issue if you use bare linux, you are free to setup any kind of software including various vpn servers.
+- Again, depending on your firmware you might be able to set up quite advanced configurations for your home network. E.g. putting all IOT devices on a separate network that does not have access to the internet. With bare linux, you are free to set up the networking rules however you like.
+- In contrast to a normal linux distro/OpenWRT with nix putting the configuration into VCS is a no-brainer.
+- last but not least I wanted to learn more about networking and these kind of things :)
+
+However, we cannot install a regular linux distribution like NixOS on an arbitrary router device and my router was no exception to that.
+Some distribution are very small and they can be installed almost everywhere. One such distribution is [OpenWRT](https://openwrt.org/) which is tailored for networking devices.
+In our case, we will need something more powerful, something that is more of a general purpose, like Raspberry PI, but tailored more for networking operations.
+Luckily, Sinovoip, a company that is known from building such boards, recently started selling their newest [Banana PI R3](https://wiki.banana-pi.org/Banana_Pi_BPI-R3)
+board (bpir3 for short) that had everything that I needed.
+
+The board has a MediaTek MT7986(Filogic 830) Quad core ARM A53 processor and 2 GB of RAM which is the minimum required for building the NixOS system.
```

Comment:
FWIW, you can run NixOS even on systems that don't have enough power to build NixOS by deploying from another system. Cross-compiled and emulated builds come in handy here too.

Reply:
Good idea, I should mention here cross-compilation and provide an example of how to set it up.